### ISSUE TITLE
Support for OpenType fonts with string-keyed CFF outlines

### ISSUE THREAD

**Issue by Selion05** [MatthiasValvekens/pyHanko#66]
Hey :) 

I'm trying to make a script so I can sign my pdfs. As far as I understand from the docs I need a stamp to make it visually obvious that that file is signed. 

I assume this is because of `%(signer)s` in the default stamp-style?
Did I mess something up creating my identity file? 

cheers Leo

**To Reproduce**
```sh
openssl req -x509 -newkey rsa:2048 -keyout mykey.pem -out cert.pem -days 365
openssl pkcs12 -export -out identity.p12 -inkey mykey.pem -in cert.pem
pyhanko sign addsig --field Sig1 pkcs12 test.pdf test-sig.pdf identity.p12
pyhanko --config pyhanko.yaml stamp --style-name default --page 1 test-sig.pdf test-sig-stamp.pdf 50 100
2021-12-11 10:33:30,710 - pyhanko.pdf_utils.font.opentype - WARNING - No ROS metadata. Is this really a CIDFont?
2021-12-11 10:33:30,718 - pyhanko.cli - ERROR - Generic processing error.
Traceback (most recent call last):
  File "pyhanko/cli.py", line 80, in pyhanko_exception_manager
    yield
  File "pyhanko/cli.py", line 1247, in stamp
    text_stamp_file(
  File "pyhanko/stamp.py", line 810, in text_stamp_file
    _stamp_file(
  File "pyhanko/stamp.py", line 784, in _stamp_file
    stamp.apply(dest_page, x, y)
  File "pyhanko/stamp.py", line 493, in apply
    stamp_ref = self.register()
  File "pyhanko/stamp.py", line 473, in register
    form_xobj = self.as_form_xobject()
  File "pyhanko/pdf_utils/content.py", line 231, in as_form_xobject
    command_stream = self.render()
  File "pyhanko/stamp.py", line 439, in render
    inner_content = self._render_inner_content()
  File "pyhanko/stamp.py", line 600, in _render_inner_content
    = self._inner_layout_natural_size()
  File "pyhanko/stamp.py", line 587, in _inner_layout_natural_size
    text_commands = self._text_layout()
  File "pyhanko/stamp.py", line 578, in _text_layout
    text = self.style.stamp_text % _text_params
KeyError: 'signer'
Error: Generic processing error.
```

```yml
stamp-styles:
    default:
        type: text
        background: __stamp__
        stamp-text: "Signed by %(signer)s\nTimestamp: %(ts)s"
        text-box-style:
            font: /usr/share/fonts/gnu-free/FreeMono.otf
    noto-qr:
        type: qr
        background: background.png
        stamp-text: "Signed by %(signer)s\nTimestamp: %(ts)s\n%(url)s"
        text-box-style:
            font: NotoSerif-Regular.otf
            leading: 13
```

**Environment**
```
linux 5.15.7-arch1-1
pyHanko, version 0.10.0
Python 3.9.9
```

**Comment by MatthiasValvekens** [comment]
Hi Leo,

Thanks for checking in!

Yeah. While the error message could be more friendly (I'll do something about that), that's indeed the cause of the problem: when stamping, there's no signer around. I'll also go ahead and point out that using the `stamp` command on a signed PDF _will_ cause any signature validator worth their salt to reject the signature... ;)  I'm not sure what you're trying to do, but you probably want to switch the order of operations there.

PS: I'm also seeing a warning in the log output about the lack of ROS data in the font you're using. It might be an OTF font with string-keyed CFF data. I'm not sure my OpenType code handles those correctly... Let me know if you encounter issues with text rendering, then at least I know that the FreeMono family is a good test-bed.

**Comment by Selion05** [comment]
Switching the order makes sense :)
I just want to sign a pdf with a visual notice in the pdf that it is signed :)  Is there an easier way then my approach? 

changing `stamp-text: "Signed by Leo\nTimestamp: %(ts)s"` yields a new error which seems related to the font warning. 

Do you know where I can find a font which works? i tried various from `fc-list` but none of them have those integer keys

```shell
> pyhanko --config pyhanko.yaml stamp --style-name default --page 1 test.pdf test-stamp.pdf 50 100
2021-12-11 11:34:27,035 - pyhanko.pdf_utils.font.opentype - WARNING - No ROS metadata. Is this really a CIDFont?
2021-12-11 11:34:27,043 - pyhanko.cli - ERROR - Generic processing error.
Traceback (most recent call last):
  File "pyhanko/cli.py", line 80, in pyhanko_exception_manager
    yield
  File "pyhanko/cli.py", line 1247, in stamp
    text_stamp_file(
  File "pyhanko/stamp.py", line 810, in text_stamp_file
    _stamp_file(
  File "pyhanko/stamp.py", line 784, in _stamp_file
    stamp.apply(dest_page, x, y)
  File "pyhanko/stamp.py", line 493, in apply
    stamp_ref = self.register()
  File "pyhanko/stamp.py", line 473, in register
    form_xobj = self.as_form_xobject()
  File "pyhanko/pdf_utils/content.py", line 231, in as_form_xobject
    command_stream = self.render()
  File "pyhanko/stamp.py", line 439, in render
    inner_content = self._render_inner_content()
  File "pyhanko/stamp.py", line 600, in _render_inner_content
    = self._inner_layout_natural_size()
  File "pyhanko/stamp.py", line 587, in _inner_layout_natural_size
    text_commands = self._text_layout()
  File "pyhanko/stamp.py", line 579, in _text_layout
    tb.content = text
  File "pyhanko/pdf_utils/text.py", line 163, in content
    wrapped_line, extent = self.put_string_line(line)
  File "pyhanko/pdf_utils/text.py", line 105, in put_string_line
    shape_result = font_engine.shape(txt)
  File "pyhanko/pdf_utils/font/opentype.py", line 449, in shape
    text_ops, (total_x, total_y) = _format_cid_glyphline_from_buffer(
  File "pyhanko/pdf_utils/font/opentype.py", line 103, in _format_cid_glyphline_from_buffer
    _format_simple_glyphline_from_buffer(buf, cid_width_callback)
  File "pyhanko/pdf_utils/font/opentype.py", line 67, in _format_simple_glyphline_from_buffer
    current_cid, width = cid_width_callback(info.codepoint)
  File "pyhanko/pdf_utils/font/opentype.py", line 413, in _get_cid_and_width
    current_cid = int(cid_str[3:])
ValueError: invalid literal for int() with base 10: ''
Error: Generic processing error.
```

**Comment by MatthiasValvekens** [comment]
>  I just want to sign a pdf with a visual notice in the pdf that it is signed :) Is there an easier way then my approach?

Yes, using a [visible signature field](https://pyhanko.readthedocs.io/en/latest/cli-guide/signing.html#creating-signature-fields) is the canonical way to do that (note: the syntax in the docs is explained for `addfields`, but also works with `addsig --field`, as noted further down).

> changing `stamp-text: "Signed by Leo\nTimestamp: %(ts)s"` yields a new error which seems related to the font warning.

Ah, that confirms my suspicion. I'll look into supporting string-keyed OTF-with-CFF fonts, then (and change the issue title accordingly). Thanks!

> Do you know where I can find a font which works? i tried various from fc-list but none of them have those integer keys

CID-keyed OTF-with-CFF and OpenType/Truetype fonts with `glyf` outlines (instead of a `CFF` table) should work out of the box. AFAIK, the Google Noto fonts all have this property: https://fonts.google.com/noto/fonts.

You can also leave off the font setting entirely, in which case pyHanko will default to an unembedded monospace font. Doesn't look very pretty, but it's unlikely to fail.

**Comment by MatthiasValvekens** [comment]
Support for non-CID-keyed CFF implemented in 8f3042d. I went with a rather hamfisted subsetting strategy (preserve all GIDs) which is a little slow and tends to produce slightly bloated subset CFF tables, but it works. From looking at some XeTeX output, I know it _can_ be done more efficiently, and I have some idea of what it's doing differently, but I'm not sure how I can get `fontTools`'s CFF subsetter to do the same. But if you can live with suboptimal performance and a couple tens of KB in file size bloat per font, that should do it. :)
